File: Domain/Entity/Message.cs
using Domain.Entity.Id;

namespace Domain.Entity;

public sealed class Message
{
    public required MessageId Id { get; set; }
    
    public required Message? PreviousMessage { get; set; }

    public required Role Role { get; init; }

    public required string Content { get; set; }

    public required DateTime CreatedUtc { get; init; }

    public required Usage? Usage { get; set; }

    public bool Visible { get; init; } = true;

    public DateTime? CompletedUtc { get; set; }
}

File: Frontend/src/types/dto/conversation/message.ts
import type { Role } from "./role";

export type Usage = {
    provider: string;
    model: string;
    tokens: number;
}

export type Message = {
    id: string;
    previousMessageId: string|null;
    role: Role;
    content: string;
    complete: boolean;
    createdUtc: Date;
    completedUtc: Date;
    usage: Usage|null;
    visible: boolean;
}